tests/Collection/CodecCollectionFunctionalTest.php

```diff
@@ -0,0 +1,530 @@
+            ],
+            'options' => [],
+        ];
+
+        yield 'No codec' => [
+            'expected' => [
+                self::createObjectFixtureResult(1, true),
+                self::createObjectFixtureResult(2, true),
+                self::createObjectFixtureResult(3, true),
+            ],
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => [
+                Document::fromPHP(self::createObjectFixtureResult(1, true)),
+                Document::fromPHP(self::createObjectFixtureResult(2, true)),
+                Document::fromPHP(self::createObjectFixtureResult(3, true)),
+            ],
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideInsertManyOptions */
+    public function testInsertMany($expected, $options): void
+    {
+        $documents = [
+            TestObject::createForFixture(1),
+            TestObject::createForFixture(2),
+            TestObject::createForFixture(3),
+        ];
+
+        $result = $this->collection->insertMany($documents, $options);
+        $this->assertSame(3, $result->getInsertedCount());
+
+        foreach ($expected as $index => &$expectedDocument) {
+            if ($expectedDocument instanceof BSONDocument && $expectedDocument->_id === null) {
+                $expectedDocument->_id = $result->getInsertedIds()[$index];
+            }
+
+            if ($expectedDocument instanceof Document && $expectedDocument->get('_id') === null) {
+                $inserted = $expectedDocument->toPHP();
+                $inserted->_id = $result->getInsertedIds()[$index];
+                $expectedDocument = Document::fromPHP($inserted);
+            }
+        }
+
+        $this->assertEquals($expected, $this->collection->find([], $options)->toArray());
+    }
+
+    public static function provideInsertOneOptions(): Generator
+    {
+        yield 'Default codec' => [
+            'expected' => TestObject::createDecodedForFixture(1),
+            'options' => [],
+        ];
+
+        yield 'No codec' => [
+            'expected' => self::createObjectFixtureResult(1, true),
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => Document::fromPHP(self::createObjectFixtureResult(1, true)),
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideInsertOneOptions */
+    public function testInsertOne($expected, $options): void
+    {
+        $result = $this->collection->insertOne(TestObject::createForFixture(1), $options);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        if ($expected instanceof BSONDocument && $expected->_id === null) {
+            $expected->_id = $result->getInsertedId();
+        }
+
+        if ($expected instanceof Document && $expected->get('_id') === null) {
```

Comment:
Are these code paths relevant? Looking at the data provider the `_id` is always set. Either way, this appears to assume that `_id` at least always _exists_ on the document, since `get()` would throw otherwise.

Reply:
Those are relevant for the "No codec" data set, as the resulting BSON document will not have an "_id" field. I've added a clarifying comment here and in the other test.

Reply:
The comment still suggests that the `_id` field won't exist ("will not have an _id field"). I think it'd make sense be clear and say that the data provider specifies a `null` value to indicate that the `_id` is unset and should be generated.

Reply:
I updated the comment - I hope this explains it. FWIW, this logic is found in a few other tests where we rely on automatically generated identifiers.